cosmwasm/contracts/crosschain-registry/src/execute.rs

```diff
@@ -295,36 +389,147 @@ pub fn chain_to_prefix_operations(
                     .clone()
                     .add_attribute("remove_chain_to_prefix", chain_name);
             }
+            FullOperation::Enable => {
+                let map_entry = CHAIN_TO_BECH32_PREFIX_MAP
+                    .load(deps.storage, &chain_name)
+                    .map_err(|_| RegistryError::AliasDoesNotExist {
+                        alias: chain_name.clone(),
+                    })?;
+                CHAIN_TO_BECH32_PREFIX_MAP.save(deps.storage, &chain_name, &(map_entry.0, true))?;
+                response
+                    .clone()
+                    .add_attribute("enable_chain_to_prefix", chain_name);
+            }
+            FullOperation::Disable => {
+                let map_entry = CHAIN_TO_BECH32_PREFIX_MAP
+                    .load(deps.storage, &chain_name)
+                    .map_err(|_| RegistryError::AliasDoesNotExist {
+                        alias: chain_name.clone(),
+                    })?;
+                CHAIN_TO_BECH32_PREFIX_MAP.save(
+                    deps.storage,
+                    &chain_name,
+                    &(map_entry.0, false),
+                )?;
+                response
+                    .clone()
+                    .add_attribute("disable_chain_to_prefix", chain_name);
+            }
         }
     }
     Ok(response)
 }
 
-// Queries
+// Struct for input data for a single chain to authorized address operation
+#[cw_serde]
+pub struct AuthorizedAddressInput {
+    pub operation: Operation,
+    pub source_chain: String,
+    pub addr: Addr,
+    pub permission: Option<Permission>,
+    pub new_addr: Option<Addr>,
+}
+
+#[cw_serde]
+pub enum Permission {
+    GlobalAdmin,
+    ChainAdmin,
+    ChainMaintainer,
+}
+
+fn permission_to_map(permission: &Permission) -> &Map<&str, Addr> {
+    match permission {
+        Permission::GlobalAdmin => &GLOBAL_ADMIN_MAP,
+        Permission::ChainAdmin => &CHAIN_ADMIN_MAP,
+        Permission::ChainMaintainer => &CHAIN_MAINTAINER_MAP,
+    }
+}
+
+pub fn authorized_address_operations(
+    deps: DepsMut,
+    sender: Addr,
+    operation: Vec<AuthorizedAddressInput>,
+) -> Result<Response, ContractError> {
+    let response = Response::new();
+    for operation in operation {
+        let addr = operation.addr;
+        let source_chain = operation.source_chain.to_lowercase();
+        let input_permission = operation.permission.unwrap();
+
+        // Check if the sender is authorized to make changes to the map of addresses authorized for the given permission
+        // GlobalAdmins can add addresses to any map
+        // ChainAdmins can modify the ChainAdmin and ChainMaintainer for their own chain
+        // ChainMaintainers can only modify the ChainMaintainer for their own chain
+        let max_permission =
+            check_is_authorized(deps.as_ref(), sender.clone(), Some(source_chain.clone()))?;
+        check_permission(input_permission.clone(), max_permission)?;
```

Comment:
I may be wrong, but is there a risk of privilege escalation here from ChainMaintainer to ChainAdmin? 

If a user is the admin for chain A but only maintainer for chain B. Max permission for that user is "admin". When checking the permissions, we don't validate that the max permission is for the chain we're affecting.

Reply:
I think you are correct, let me add a test case that proves you are right and patch this. Nice find :) (I didn't consider a single address being used across multiple chains when writing this)

Reply:
After creating a test case for this, I found out that the risk of privilege escalation is not present. This is because we pull the permission level via `check_is_authorized` for a specific chain. Therefore, a sender that is a chain_admin for chainA but only maintainer for chainB would fail the `check_is_chain_admin` for chainB but would pass the `check_is_chain_maintainer` check and return a maximum permission of ChainMaintainer for whatever that call is.

Regaurdless, you had me convinced initially and I'm glad we now have a test case for this. Thanks!

Reply:
https://github.com/osmosis-labs/osmosis/pull/4553/commits/d2afa3baab439f214c1b7e8cdde201a0cf492c28